src/instructlab/utils.py

```diff
@@ -556,15 +566,89 @@ def split_hostport(hostport: str) -> tuple[str, int]:
     return hostname, port
 
 
-def convert_messages_to_legacy_dataset(
+def is_pretraining_dataset(ds: List[MessageSample]) -> bool:
+    """
+    Determines whether or not the given dataset is the phase007 pretraining
+    dataset.
+    """
+    if not ds:
+        return False
+    sample = ds[0]
+    return any(m["role"] == "pretraining" for m in sample["messages"])
+
+
+def get_user_assistant_from_pretraining(pretraining: str) -> Tuple[str, str]:
+    """
+    Given a pretraining message sample which contains the tokens '<|user|>' and '<|assistant|>',
+    return the strings corresponding to the user & assistant inputs.
+
+    This function assumes that a given message begins with the '<|user|>' token
+    and contains '<|assistant|>' somewhere later.
+
+    Raises ValueError if neither the '<|user|>' or '<|assistant|>' tokens exist
+    within the given sample.
+    """
+    if SDGTokens.ASSISTANT not in pretraining:
+        raise ValueError(
+            f"pretraining sample doesn't contain the '{SDGTokens.ASSISTANT}' token: {pretraining}"
+        )
+    if SDGTokens.USER not in pretraining:
+        raise ValueError(
+            f"pretraining sample doesn't contain the '{SDGTokens.USER}' token: {pretraining}"
+        )
+    user, assistant = pretraining.split(SDGTokens.ASSISTANT)
+    user = user.removeprefix(SDGTokens.USER)
+    return user, assistant
+
+
+def convert_pretraining_messages_to_legacy_dataset(
     dataset: List[MessageSample],
 ) -> List[LegacyMessageSample]:
     """
-    Converts the new HuggingFace messages dataset format to the legacy format.
-    For reference, see: https://huggingface.co/docs/transformers/en/chat_templating#templates-for-chat-modelos
+    Given a Phase007 pretraining dataset that's in the messages format, returns
+    a version that's been converted to be compatible with the legacy ilab data format.
 
-    **Note**: The legacy dataset format assumes only a turn of 1. All extra turns will be dropped.
-    This means that we only look at the first 3 messages, and everything afterwards will be ignored.
+    This function assumes that each sample contains at least a single message with the
+    "pretraining" role. If a "system" role exists then it's also parsed out, otherwise
+    an empty string is set for the system.
+    """
+    converted_dataset: List[LegacyMessageSample] = []
+    for sample in dataset:
+        pretraining = next(
+            (
+                msg["content"]
+                for msg in sample["messages"]
+                if msg["role"] == "pretraining"
+            ),
+            None,
+        )
+        # not sure if 'system' will exist in phase007 or not - leaving here just in case
+        system = next(
+            (msg["content"] for msg in sample["messages"] if msg["role"] == "system"),
+            "",
+        )
+        if not pretraining:
```

Comment:
nit: I think you could raise it before `system` is calculated.

Reply:
Good point.